### ISSUE TITLE
tf.train.L_BFGS_Optimizer

### ISSUE THREAD

**Issue by zfrenchee** [tensorflow/tensorflow#446]
This would be a great addition to tensorflow, and is conspicuously missing. Is there some specific reason it's missing, or is it in the works?

**Comment by martinwicke** [comment]
It is not currently in the works, and it would be a great addition to TensorFlow. Any contribution in this area would be very welcome.

**Comment by joshburkart** [comment]
Putting in a full implementation of L-BFGS seems like a nontrivial task, but a first step could be introducing some sort of line search functionality, which could be an option to `GradientDescentOptimizer` as well when in full-batch (no-dropout) mode?

Perhaps there could be e.g. a `FullBatchOptimizer` base class that implements line searches, with a virtual function `StepDirection` to compute line search directions (which could then be implemented separately for gradient descent, conjugate gradient, L-BFGS, etc.)?

Thoughts?

**Comment by smcantab** [comment]
I agree that it would be a great addition and I would be happy to contribute. I have been using something analogous to this slick implementation [lbfgs_cpp](https://github.com/js850/lbfgs_cpp) by @js850.
For instance the [`lbfgs_cpp::compute_lbfgs_step`](https://github.com/js850/lbfgs_cpp/blob/master/lbfgs.cpp#L165) would fit within your proposed `StepDirection` virtual function, although it would have to be rewritten to support gpus. Likewise for the backtracking linesearch which one might want to use on top of it.

**Comment by ogrisel** [comment]
Mark Schmidt's [minFunc](https://www.cs.ubc.ca/~schmidtm/Software/minFunc.html) is famous for its state of the art implementation of line searches. It's probably a very good reference implementation to benchmark against.

As a side note, the implementation of l-BFGS-b in scipy is a wrapper for the official FORTRAN implementation:

http://docs.scipy.org/doc/scipy-0.17.0/reference/generated/scipy.optimize.fmin_l_bfgs_b.html
http://users.iems.northwestern.edu/~nocedal/lbfgsb.html

which is probably another good reference implementation to benchmark against.

**Comment by yaroslavvb** [comment]
BTW, scipy lBFGS seems to be slower than Matlab's minFunc
I tried running https://github.com/jatinshah/ufldl_tutorial, which uses scipy's lBFGS, and it felt an order of magnitude slower than equivalent Matlab version

**Comment by ogrisel** [comment]
@yaroslavvb you have to be careful to use equivalent stopping criterion which is not always easy to do.

**Comment by darrengarvey** [comment]
I had a question to the authors on testing an implementation of this.

For the other optimisers like Adam there is a python function that does the equivalent optimisation and then the test checks that the real implementation comes up with the same result. That seems reasonable given how simple the other optimisers are, but what about something more involved like L-BFGS-B?

The implementation could be compared to what scipy gets, but then again bringing in scipy as a dependency just for the sake of one test seemed, well, excessive. A full Python implementation just for the tests seems brittle as the Python implementation could just as easily be buggy. Perhaps a pure python implementation is the way to go?

**Comment by joshburkart** [comment]
We have an external optimizer interface module coming to contrib/ that should address this. Allows controlling a TF session using e.g. scipy.optimize.minimize(). Should be available soon!

**Comment by darrengarvey** [comment]
@joshburkart - Sounds like exactly what I was looking for!

**Comment by joshburkart** [comment]
This is done now and should be closed. :) See [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/opt/python/training/external_optimizer.py#L235) for usage instructions.

**Comment by yaroslavvb** [comment]
Do you have any numbers on the speed? I've been working on a GPU version of l-BFGS optimizer in tensorflow, so it would be interesting to compare

**Comment by joshburkart** [comment]
Hmm nope no numbers. It involves moving data not only off the GPU but also into Python for every loss or gradient evaluation, so I imagine you wouldn't want to use it for small optimization problems if efficiency is an issue. But as your loss becomes more expensive to evaluate this disadvantage becomes less important.

**Comment by yuanbyu** [comment]
Could persistent tensors be used to avoid the transfers from GPU to python?  You may want to take a look at python/ops/session_ops.py and python/kernel_tests/session_ops_test.py.

**Comment by joshburkart** [comment]
@yuanbyu Not if we're using a Scipy optimizer to control a TF session. That's what we have working now. Keeping everything in TF the whole time would require a custom TF implementation.

**Comment by yuanbyu** [comment]
I guess that the Scipy optimizer makes a sequence of run calls in a TF session? Are there cases that a run call produces a tensor but the tensor is only used in a subsequent run call?  Persistent tensor would make it possible to keep such tensors "in-place" on the device that produced them, avoid moving them back and forth between the python client and the device.

On the other hand, it would be really nice to have a real native TF implementation, possibly using a nested while loop.

**Comment by yaroslavvb** [comment]
So I have [a version](https://github.com/yaroslavvb/lbfgs/blob/master/lbfgs_optimize_reference.py) of l-BFGS that relies on persistent tensors to keep data in TensorFlow runtime, and [Immediate mode](https://github.com/tensorflow/tensorflow/pull/2747) for syntactic sugar. It's a line-by-line port of [lbfgs.lua](https://github.com/torch/optim/blob/master/lbfgs.lua) -- here's [a comparison](https://docs.google.com/spreadsheets/d/17RwQSIV6E0uXSATec2-Lu_GdhoqEsbabFEWHg1k5wzc/edit#gid=0) . I checked numerics on some toy problems and it tracks Lua implementation [almost exactly](https://github.com/yaroslavvb/lbfgs/blob/master/Porting%20Torch%20l-BFGS%20to%20immediate.ipynb) .

Persistent tensor operations adds 200-300 usec overhead for each TensorFlow operation, so if "the meat" of each tensor operation is below 300usec, then a fully native TensorFlow implementation will be more efficient.

**Comment by weiliu620** [comment]
@yaroslavvb Any plan about when to add your implementation into official TensorFlow code?

**Comment by yaroslavvb** [comment]
@weiliu620 It will not go into official TensorFlow code. There's considerable complexity in the module and nobody in the core team feels comfortable enough to assume ownership of it. However, since it's pure Python, it doesn't need to be part of the core. I'm working on getting a single-file version working so you could "install" the mode by copy pasting it into your file. The main core dependency is persistent tensors which should be working in upcoming release (0.10) after which I can release this mode. Do you have any specific use-cases in mind? In particular, I have the regular l-BFGS, but wondering if I need to do specific line-search additions

**Comment by weiliu620** [comment]
@yaroslavvb While I'm testing the neural-style code by @jcjohnson, I found using lbfgs optimizer got cost function ~12% less than using adam optimizer, with fixed number of iterations. There are also some [discussions ](https://www.reddit.com/r/MachineLearning/comments/3u1ssk/neural_style_in_tensorflow/) on these two optimizers on style transfer. 

So that is the use case that I run into this issue. Thanks for the work!

**Comment by weiliu620** [comment]
@joshburkart Can you provide a minimal example of using the loss_callback? The in-code examples are great, just not using loss_callback. Thanks!

**Comment by joshburkart** [comment]
Example callback:
```python
def callback(x):
  # x is the current packed set of values being optimized.
  print(np.linalg.norm(x))
```
It would be nice if the callback received e.g. a map from tensor name to current value, rather than the internal packed representation as a single vector, but that's currently not implemented. Feel free to work on it!

Note that we're deprecating `grad_callback` (we're going to run both loss and loss grad in one `session.run` call) so best not to use that.

**Comment by joshburkart** [comment]
Whoa! Please ignore my previous completely incorrect post. Forgot that I actually did implement this correctly the first time. :P The callback does in fact get arbitrary fetches as appropriate-shaped arrays. So here's an example where the callback receives the loss itself (but it could receive any other tensor in the graph):

```python
loss = ...

i = 0
def callback(loss):
  nonlocal i
  print('Loss evaluation #', i, ', loss:', loss)
  i += 1

scipy_optimizer =  tf.contrib.opt.ScipyOptimizerInterface(loss)
scipy_optimizer.minimize(session, fetches=[loss], loss_callback=callback)
```

I should really rename `fetches` to `callback_fetches`, since this is confusing...

**Comment by dzhwinter** [comment]
Is there any progress on this enhancement? I'm working on that and need a comparison.

**Comment by weiliu620** [comment]
@joshburkart tested and the callback works. Thanks for the great work!

**Comment by olegaulov** [comment]
Hi @joshburkart and @weiliu620, would you mid elaborating a bit further on how to use a BFGS optimizer? I'm a bit too novice to fill in the missing pieces from Josh's example. I have a basic running net that uses the following:

`loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(trn_logits, tf_train_labels))
optimizer = tf.train.AdamOptimizer(0.05).minimize(loss)
`
I'm a bit confused as to how to deal with the: session, fetches=[loss], loss_callback=callback part.

If you have suggestions or some pointers to where to look for answers I'd really appreciate it.
Thanks!

**Comment by joshburkart** [comment]
@olegaulov There are some examples in the [docstring](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/opt/python/training/external_optimizer.py#L246) :)

**Comment by olegaulov** [comment]
Thanks @joshburkart ! Just what I was looking for.

**Comment by tobegit3hub** [comment]
Thanks @joshburkart for your contribution. Would you mind to close this issue? @zfrenchee @martinwicke

**Comment by ncullen93** [comment]
For future lurkers, here's an end-to-end example of using this optimizer on the UFDL sparse ae tutorial

https://github.com/ncullen93/UFDL_Tensorflow/blob/master/tf_sparse_ae.py

**Comment by zhaojunz** [comment]
@ncullen93 the link is not valid anymore. does anyone have a backup copy?

**Comment by gindeleo** [comment]
@zhaojunz Yes, the link is dead. Maybe this will help on how to use a scipy optimizer (L-BFGS for example):

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/opt/python/training/external_optimizer_test.py

**Comment by flowice** [comment]
@weiliu620 @joshburkart 
I have tried the L-BFGS method in the neural-art code by @jcjohnson . 
It does work and return more stable results than Adam.
However, I find the optimization became a little slower.

Also, is there any way to probe at iteration progress (print stuff / generate checkpoint images using interval variables) like the original code does.
Since I find in the official comment doc that "variables subject to optimization are updated in-place at the end of optimization".

**Comment by rabeehk** [comment]
Hello @flowice 
Do you mind sharing the codes of the L-BFGS in tensorflow with me? So you have the working code? 
thanks a lot
Best
Rabeeh

**Comment by flowice** [comment]
@rabeehk  it is easy to use. I just follow the official doc.
The core code is here:

`
train_step = tf.contrib.opt.ScipyOptimizerInterface(
                loss,
                method='L-BFGS-B',
                options={'maxiter': iterations})

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    train_step.minimize(sess)
    
`

**Comment by rabeehk** [comment]
@flowice Thanks a lot for the help. I tried it, for my loss function, it becomes "nan", maybe the learning rate is too high, I was wondering if you may know how I can modify the learning rate for this method? I could not find this option in the codes. Thanks a lot in advance.

**Comment by debasish83** [comment]
We have added LBFGS based optimizers in Breeze which is being used in Spark MLlib for example and the C++ port from fortran is available through multiple projects (I benchmarked this for example https://github.com/chokkan/liblbfgs) but given that neural nets are based on non-convex loss (original distbelief paper also found that sgd + adagrad works better than LBFGS), is it useful to add LBFGS in tensorflow ? Most likely users are now running convex models in tensorflow and for such models LBFGS is a good fit compared to SGD + Adagrad ?

**Comment by yaroslavvb** [comment]
@debasish83 I've implemented l-BFGS in TensorFlow [here](https://medium.com/@yaroslavvb/tensorflow-meets-pytorch-with-eager-mode-714cce161e6c).

Some benchmarks are [here](https://medium.com/@yaroslavvb/tensorflow-meets-pytorch-with-eager-mode-714cce161e6c). It's not as performant as it could be, but @alextp and friends are working on optimizing the underlying framework.

One thing that should significantly speed up implementation above is to reduce Python/TensorFlow round-trips by replacing inner python loops (two-step recursion) with defuns, something like this (h/t @agarwal-ashish)

```
def f(k, q, *args):
  old_dirs = args[:k]
  ro = args[k:2*k]
  old_steps = args[2*k:3*k]
  for i in range(k-1, -1, -1):
    al = dot(old_dirs[i], q) * ro[i]
    q = q - al * old_steps[i]
  return q

@function.defun
def g(q, *args):
  return f(100, q, *args)

 q = -g
  if k == 100:
    q = g(q, *(old_dirs + ro + old_steps))
  else:
    q = f(k, q, *(old_dirs + ro + old_steps))

```

**Comment by debasish83** [comment]
@yaroslavvb If the model is distributed on a set of ps nodes, will this https://github.com/yaroslavvb/stuff/blob/master/eager_lbfgs/eager_lbfgs.py work ? I think we need to run BFGS loop on the chief node which coordinate the worker and ps nodes...distbelief paper did that...

**Comment by yaroslavvb** [comment]
Nope, there's no distributed support in Eager. So if you want to large scale implementation, you'll need to use classic execution mode.

There used to be l-BFGS implemented in standard TensorFlow but it was deleted because it never got robust enough. Because tensorflow distributed framework is quite low-level so people sometimes get surprised by how much work it takes to get something robust. Following comment/commits in sync_replicas_optimizer gives an overview of some pitfalls -- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/sync_replicas_optimizer.py

**Comment by Hoeze** [comment]
K-FAC is a second-order optimizer, isn't it?
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/kfac

**Comment by mmoller64** [comment]
Hi all, 
Have any one considered implementing the Scaled Conjugate Gradient Algorithm (SCG) in Tensorflow?
SCG avoids the line search alltogether. SCG is currently implemented in Mathlab as 'trainscg' method. 
I would like to contribute, but I am, however, currently not that familiar with Tensorflow. 
For reference see: 
- A scaled conjugate gradient algorithm for fast supervised learning, M. F. Moller, Neural Networks 6 (4), 525-533, 1993.
 and more general info on: 
- Efficient Training of Feed-Forward Neural Networks

The Hessian times a vector (used in the SCG) can be calculated exact and/or approximated. 
--
[A scaled conjugate gradient algorithm for fast supervised learning.pdf](https://github.com/tensorflow/tensorflow/files/2253652/A.scaled.conjugate.gradient.algorithm.for.fast.supervised.learning.pdf)

[Efficient training of feed-forward neural networks - moller - phd thesis.pdf](https://github.com/tensorflow/tensorflow/files/2253657/Efficient.training.of.feed-forward.neural.networks.-.moller.-.phd.thesis.pdf)

**Comment by Hoeze** [comment]
fyi, there is a BFGS-Implementation in tensorflow/probability:
https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/optimizer/bfgs.py

**Comment by cslr** [comment]
I have implemented L-BFGS optimizer in my dinrhiw2 C++ machine learning library. When implemented well line search doesn't take that much time and seem to be get results equally fast than with first order optimizers. For some problems L-BFGS can seem to be able to side-step some local minimas and get better results.

https://github.com/cslr/dinrhiw2/blob/RBM_test/src/math/LBFGS.cpp

**Comment by johnpjust** [comment]
FYI -- i've put out a request to update and make the scipyoptimizer interface compatible with eager execution.  Additionally would like to tie in an early stopping option (i list a hack-y way to do this already).  Anyone willing to help would be much appreciated.  The scipy optimizer has L-BFGS + CG + others, so it would be worth the time!

**Comment by Overdrivr** [comment]
@yaroslavvb Do you have any benchmarks againsts scipy's CPU version ? That would be very interesting.

**Comment by bright1993ff66** [comment]
@flowice Thank you very much for your answer. But could I apply the ```bfgs``` method in each batch of data and do the parameter update?

I have also checked the other package like ```neupy``` and it seems like they provide an answer. But it seems like this package relies on ```theano``` and makes it very difficult to run codes in the 64-bit computer....

**Comment by Demetrio92** [comment]
this blog posts explains how to use `tfp.optimizer.lbfgs_minimize` and provides a reproducible example. 
https://pychao.com/2019/11/02/optimize-tensorflow-keras-models-with-l-bfgs-from-tensorflow-probability/
